File: backend/src/shared/base.service.ts
import Datastore from 'nedb-promises';
import { BaseDocument } from '../types/models';

interface FindOptions {
  sort?: Record<string, number>;
  skip?: number;
  limit?: number;
}

/**
 * Generic CRUD base service using NeDB.
 * Module services extend this via composition or inheritance (OCP).
 */
export class BaseService<T extends BaseDocument> {
  protected db: Datastore;

  constructor(datastore: Datastore) {
    this.db = datastore;
  }

  async findAll(query: Record<string, unknown> = {}, options: FindOptions = {}): Promise<T[]> {
    let cursor = this.db.find<T>(query);
    if (options.sort) cursor = cursor.sort(options.sort);
    if (options.skip) cursor = cursor.skip(options.skip);
    if (options.limit) cursor = cursor.limit(options.limit);
    return cursor.exec();
  }

  async findOne(query: Record<string, unknown>): Promise<T | null> {
    return this.db.findOne<T>(query);
  }

  async findById(id: string): Promise<T | null> {
    return this.db.findOne<T>({ _id: id });
  }

  async create(data: Omit<T, '_id' | 'createdAt' | 'updatedAt'>): Promise<T> {
    const doc = { ...data, createdAt: new Date(), updatedAt: new Date() };
    return this.db.insert<T>(doc as T);
  }

  async update(id: string, data: Partial<T>): Promise<T | null> {
    const updateData = { ...data, updatedAt: new Date() };
    await this.db.update({ _id: id }, { $set: updateData });
    return this.findById(id);
  }

  async delete(id: string): Promise<number> {
    return this.db.remove({ _id: id }, {});
  }

  async count(query: Record<string, unknown> = {}): Promise<number> {
    return this.db.count(query);
  }
}

File: backend-dotnet-net22/Services/BaseService.cs
using AllTrackIn.Api.Models;
using LiteDB;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Linq.Expressions;

namespace AllTrackIn.Api.Services
{
    public class BaseService<T> where T : BaseDocument
    {
        public readonly ILiteCollection<T> Collection;

        public BaseService(ILiteCollection<T> collection)
        {
            Collection = collection;
        }

        public List<T> FindAll(Expression<Func<T, bool>> predicate = null)
            => predicate == null ? Collection.FindAll().ToList() : Collection.Find(predicate).ToList();

        public T FindOne(Expression<Func<T, bool>> predicate)
            => Collection.FindOne(predicate);

        public T FindById(string id)
            => Collection.FindById(id);

        public T Create(T document)
        {
            document.Id = ObjectId.NewObjectId().ToString();
            document.CreatedAt = DateTime.UtcNow;
            document.UpdatedAt = DateTime.UtcNow;
            Collection.Insert(document);
            return document;
        }

        public T Update(T document)
        {
            document.UpdatedAt = DateTime.UtcNow;
            Collection.Update(document);
            return document;
        }

        public bool Delete(string id)
            => Collection.Delete(id);

        public int Count(Expression<Func<T, bool>> predicate)
            => Collection.Count(predicate);
    }
}
